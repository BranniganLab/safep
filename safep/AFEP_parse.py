"""
Large datasets can be difficult to parse on a workstation due to inefficiencies 
in the way data is represented for pymbar. When possible, reduce the size of your dataset.
"""
import argparse
import os
import warnings
from dataclasses import dataclass
from typing import NamedTuple
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy as sp
from alchemlyb.parsing import namd

from scipy.constants import R, calorie

import safep

warnings.simplefilter(action="ignore", category=FutureWarning)


@dataclass
class FepRun:
    u_nk: pd.DataFrame
    perWindow: pd.DataFrame
    cumulative: pd.DataFrame
    forward: pd.DataFrame
    forward_error: pd.DataFrame
    backward: pd.DataFrame
    backward_error: pd.DataFrame
    per_lambda_convergence: pd.DataFrame
    color: str

    def __post_init__(self):
        # Run the BAR estimator on the fep data
        self.perWindow, self.cumulative = safep.do_estimation(self.u_nk)
        (
            self.forward,
            self.forward_error,
            self.backward,
            self.backward_error,
        ) = safep.do_convergence(self.u_nk)  # Used later in the convergence plot
        self.per_lambda_convergence = safep.do_per_lambda_convergence(self.u_nk)


def do_agg_data(dataax, plotax):
    """
    Aggregates data from a given matplotlib axis, computes statistical measures,
    and displays them on another axis.

    Parameters:
    dataax (matplotlib.axes.Axes): The axis containing lines with data to be aggregated.
    plotax (matplotlib.axes.Axes): The axis where the statistical summary will be displayed.

    Returns:
    matplotlib.axes.Axes: The plot axis with the statistical summary text added.
    """
    agg_data = []
    lines = dataax.lines
    for line in lines:
        agg_data.append(line.get_ydata())
    flat = np.array(agg_data).flatten()
    kernel = sp.stats.gaussian_kde(flat)
    pdf_x = np.linspace(-1, 1, 1000)
    pdf_y = kernel(pdf_x)
    std = np.std(flat)
    average = np.average(flat)
    temp = pd.Series(pdf_y, index=pdf_x)
    mode = temp.idxmax()

    textstr = (r"$\rm mode=$" + f"{np.round(mode,2)}" + "\n" + fr"$\mu$={np.round(average,2)}" +
               "\n" + fr"$\sigma$={np.round(std,2)}")
    props = {"boxstyle": 'square', "facecolor": 'white', "alpha": 1}
    plotax.text(
        0.175,
        0.95,
        textstr,
        transform=plotax.transAxes,
        fontsize=14,
        verticalalignment="top",
        bbox=props,
    )

    return plotax


class AFEPArgumentParser(argparse.ArgumentParser):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.add_argument(
            "--path",
            type=str,
            help="The absolute path to the directory containing the fepout files",
            default=".",
        )
        self.add_argument(
            "--fepoutre",
            type=str,
            help="A regular expression that matches the fepout files to be parsed.",
            default="*.fep*",
        )
        self.add_argument(
            "--replicare",
            type=str,
            help="A regular expression that matches the replica directories",
            default="Replica?",
            required=True,
        )
        self.add_argument(
            "--temperature",
            type=float,
            help="The temperature at which the FEP was run.",
            default=303.15,
        )
        self.add_argument(
            "--detectEQ",
            type=bool,
            help="Flag for automated equilibrium detection.",
            default=True,
        )
        self.add_argument(
            "--fittingMethod",
            type=str,
            help="Method for fitting the forward-backward discrepancies (hysteresis)." +
            "LS=least squares, ML=maximum likelihood Default: LS",
            default="LS",
        )
        self.add_argument(
            "--maxSize",
            type=float,
            help="Maximum total file size in GB." +
            "This is MUCH less than the required RAM. Default: 1",
            default=1,
        )
        self.add_argument(
            "--makeFigures",
            type=bool,
            help="Run additional diagnostics and save figures to the directory. default: False",
            default=0,
        )


KILO = 1000
COLORS = ["blue", "red", "green", "purple", "orange", "violet", "cyan"]


def initialize_general_figure(RT_kcal_per_mol, key, feprun):
    fig, axes = safep.plot_general(
        feprun.cumulative,
        None,
        feprun.perWindow,
        None,
        RT_kcal_per_mol,
        hysttype="lines",
        label=key,
        color=feprun.color,
    )
    axes[1].legend()

    return fig, axes

def report_number_and_size_of_fepout_files(fepoutFiles):
    total_size = 0
    for file in fepoutFiles:
        total_size += os.path.getsize(file)
    print(f"Will process {len(fepoutFiles)} fepout files." +
          f"\nTotal size:{np.round(total_size/10**9, 2)}GB")


class AFEPArguments(NamedTuple):
    dataroot: Path
    replica_pattern: str
    replicas: list[Path]
    filename_pattern: str
    temperature: float
    RT_kcal_per_mol: float
    detectEQ: bool
    makeFigures: bool

    @classmethod
    def from_AFEPArgumentParser(cls, parser: AFEPArgumentParser) -> AFEPArgumentParser:
        args = parser.parse_args()

        dataroot = Path(args.path)
        replica_pattern = args.replicare
        replicas = dataroot.glob(replica_pattern)
        filename_pattern = args.fepoutre

        temperature = args.temperature
        args.RT_kcal_per_mol = R / (KILO * calorie) * temperature
        detectEQ = args.detectEQ

        return cls(dataroot, replica_pattern, replicas, filename_pattern, temperature,
                   args.RT_kcal_per_mol, detectEQ, args.makeFigures)


def read_and_decorrelate(args, replica, unkpath, fepoutFiles):
    fig, ax = plt.subplots()

    u_nk = namd.extract_u_nk(fepoutFiles, args.temperature)
    u_nk = u_nk.sort_index(axis=0, level=1).sort_index(axis=1)
    safep.plot_samples(ax, u_nk, color="blue", label="Raw Data")

    if args.detectEQ:
        print("Detecting equilibrium")
        u_nk = safep.detect_equilibrium_u_nk(u_nk)
        safep.plot_samples(ax, u_nk, color="orange", label="Equilibrium-Detected")

    plt.savefig(args.dataroot.joinpath(f"{str(replica)}_FEP_number_of_samples.pdf"))
    safep.save_UNK(u_nk, unkpath)
    return u_nk


def process_replicas(args, itcolors):
    fepruns = {}
    for replica in args.replicas:
        print(f"Reading {replica}")
        unkpath = replica.joinpath("decorrelated.csv")
        u_nk = None
        if unkpath.is_file():
            print("Found existing dataframe. Reading.")
            u_nk = safep.read_UNK(unkpath)
        else:
            print(f"Didn't find existing dataframe at {unkpath}. Checking for raw fepout files.")
            fepoutFiles = list(replica.glob(args.filename_pattern))
            report_number_and_size_of_fepout_files(fepoutFiles)

            if len(list(fepoutFiles)) > 0:
                print("Reading fepout files")
                u_nk = read_and_decorrelate(args, replica, unkpath, fepoutFiles)
            else:
                print(f"WARNING: no fepout files found for {replica}. Skipping.")

        if u_nk is not None:
            fepruns[str(replica)] = FepRun(u_nk, None, None, None, None, None, None, None,
                                           next(itcolors))
    return fepruns


def add_to_general_figure(fig, axes, args, key, feprun):
    fig, axes = safep.plot_general(
                    feprun.cumulative,
                    None,
                    feprun.perWindow,
                    None,
                    args.RT_kcal_per_mol,
                    fig=fig,
                    axes=axes,
                    hysttype="lines",
                    label=key,
                    color=feprun.color,
                )
    
    return fig, axes

def add_summary_stats(do_agg_data, mean, sterr, axes):
    axes[3] = do_agg_data(axes[2], axes[3])

    axes[0].set_title(str(mean) + r"$\pm$" + str(sterr) + " kcal/mol")
    axes[0].legend()
    return axes

def do_shared_convergence_plot(args, fepruns, dGs):
    fig, convAx = plt.subplots(1, 1)

    for key, feprun in fepruns.items():
        convAx = safep.convergence_plot(
                convAx,
                feprun.forward * args.RT_kcal_per_mol,
                feprun.forward_error * args.RT_kcal_per_mol,
                feprun.backward * args.RT_kcal_per_mol,
                feprun.backward_error * args.RT_kcal_per_mol,
                fwd_color=feprun.color,
                bwd_color=feprun.color,
                errorbars=False,
            )
        convAx.get_legend().remove()

    (forward_line,) = convAx.plot([], [],
                                      linestyle="-",
                                      color="black",
                                      label="Forward Time Sampling")
    (backward_line,) = convAx.plot([], [],
                                       linestyle="--",
                                       color="black",
                                       label="Backward Time Sampling")
    convAx.legend(handles=[forward_line, backward_line])
    ymin = np.min(dGs) - 1
    ymax = np.max(dGs) + 1
    convAx.set_ylim((ymin, ymax))

    return fig, convAx

def do_per_lambda_convergence_shared_axes(do_agg_data, initialize_general_figure, args, fepruns, mean, sterr, axes):
    genfig = None
    for key, feprun in fepruns.items():
        if genfig is None:
            genfig, genaxes = initialize_general_figure(args.RT_kcal_per_mol, key, feprun)
        else:
            genfig, genaxes = safep.plot_general(
                    feprun.cumulative,
                    None,
                    feprun.perWindow,
                    None,
                    args.RT_kcal_per_mol,
                    fig=genfig,
                    axes=genaxes,
                    hysttype="lines",
                    label=key,
                    color=feprun.color,
                )
    plt.delaxes(genaxes[0])
    plt.delaxes(genaxes[1])

    genaxes[3] = do_agg_data(axes[2], axes[3])
    genaxes[2].set_title(str(mean) + r"$\pm$" + str(sterr) + " kcal/mol")

    for txt in genfig.texts:
        print(1)
        txt.set_visible(False)
        txt.set_text("")

    return genfig, genaxes

if __name__ == "__main__":
    parser = AFEPArgumentParser()
    args = AFEPArguments.from_AFEPArgumentParser(parser)

    itcolors = iter(COLORS)

    # # Extract key features from the MBAR fitting and get ΔG
    # Note: alchemlyb operates in units of kT by default.
    # We multiply by RT to convert to units of kcal/mol.

    fepruns = process_replicas(args, itcolors)

    toprint = ""
    dGs = []
    errors = []
    for key, feprun in fepruns.items():
        cumulative = feprun.cumulative
        dG = np.round(cumulative.BAR.f.iloc[-1] * args.RT_kcal_per_mol, 1)
        error = np.round(cumulative.BAR.errors.iloc[-1] * args.RT_kcal_per_mol, 1)
        dGs.append(dG)
        errors.append(error)

        changeAndError = f"{key}: \u0394G = {dG}\u00B1{error} kcal/mol\n"
        toprint += changeAndError

    toprint += "\n"
    mean = np.average(dGs)

    # If there are only a few replicas,
    # the MBAR estimated error will be more reliable, albeit underestimated
    if len(dGs) < 3:
        sterr = np.sqrt(np.sum(np.square(errors)))
    else:
        sterr = np.round(np.std(dGs), 1)
    toprint += f"mean: {mean} kcal/mol\n" + f"sterr: {sterr} kcal/mol"
    print(toprint)

    if args.makeFigures == 1:
        fig = None
        for key, feprun in fepruns.items():
            if fig is None:
                fig, axes = initialize_general_figure(args.RT_kcal_per_mol, key, feprun)
            else:
                fig, axes = add_to_general_figure(fig, axes, args, key, feprun)

        # hack to get aggregate data:
        axes = add_summary_stats(do_agg_data, mean, sterr, axes)
        fig.savefig(args.dataroot.joinpath("FEP_general_figures.pdf"))

        # # Plot the estimated total change in free energy as a function of simulation time;
        # contiguous subsets starting at t=0 ("Forward") and t=end ("Reverse")

        fig, convAx = do_shared_convergence_plot(args, fepruns, dGs)
        fig.savefig(args.dataroot.joinpath("FEP_convergence.pdf"))

        fig, axes = do_per_lambda_convergence_shared_axes(do_agg_data, initialize_general_figure, args, fepruns, mean, sterr, axes)
        fig.savefig(args.dataroot.joinpath("FEP_perLambda_convergence.pdf"))

        plt.show()
