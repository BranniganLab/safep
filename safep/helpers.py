# Import block
import numpy as np

from scipy.stats import norm

# Calculate the coefficient of determination:
def get_Rsq(X, Y, Yexpected):
    '''
    Calculate the coefficient of determination for arbitrary fits.
    Arguments: X (inputs), Y (experimental data), Yexpected (fitted or predicted data)
    '''
    residuals = Y-Yexpected
    SSres = np.sum(residuals**2)
    SStot = np.sum((X-np.mean(X))**2)
    R = 1-SSres/SStot
    return R
 
#Wrappers
def cum_Fn(x, m, s):
    '''
    Wrapper for the normal cumulative density function
    '''
    r = norm.cdf(x, m, s)
    return r

def pdf_Fn(x,m,s):
    '''
    Wrapper for the normal probability density function
    '''
    r = norm.pdf(x,m,s)
    return r

def moving_average(x, w):
    '''
    Generate a moving average over x with window width w
    Arguments: x (data), w (window width)
    Returns: a moving window average
    '''
    return np.convolve(x, np.ones(w), 'same') / w