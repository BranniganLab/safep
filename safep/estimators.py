# Import block
import matplotlib.pyplot as plt

import numpy as np
from numpy.lib.stride_tricks import sliding_window_view

from scipy.stats import linregress as lr
from scipy.stats import norm
from scipy.special import erfc
from scipy.optimize import curve_fit as scipyFit
from scipy.stats import skew

import pandas as pd

from alchemlyb.visualisation.dF_state import plot_dF_state
from alchemlyb.parsing import namd
from alchemlyb.estimators import BAR
from alchemlyb.visualisation.dF_state import plot_dF_state
from alchemlyb.visualisation import plot_convergence

import re
from tqdm import tqdm #for progress bars
from natsort import natsorted #for sorting "naturally" instead of alphabetically
from glob import glob #file regexes

from .helpers import *


def get_exponential(u_nk):
    '''
    Get exponential estimation of the change in free energy.
    Arguments: u_nk in alchemlyb format
    Returns: l[ambdas], l_mid [lambda window midpoints], dG_f [forward estimates], dG_b [backward estimates]
    ''' 

    groups = u_nk.groupby(level=1)
    dG=pd.DataFrame([])
    for name, group in groups:
        dG[name] = np.log(np.mean(np.exp(-1*group), axis=0))

    dG_f=np.diag(dG, k=1)
    dG_b=np.diag(dG, k=-1)

    l=dG.columns.to_list()
    l_mid = np.mean([l[1:],l[:-1]], axis=0)

    return l, l_mid, dG_f, dG_b
    
def get_BAR(bar):
    '''
    Extract key information from an alchemlyb.BAR object. Useful for plotting.
    Arguments: a fitted BAR object
    Returns: l[ambdas], l_mid [lambda window midpoints], f [the cumulative free energy], df [the per-window free energy], ddf [the per-window errors], errors [the cumulative error]
    '''
    
    states = bar.states_

    f = bar.delta_f_.iloc[0,:] # dataframe
    l = np.array([float(s) for s in states])
    # lambda midpoints for each window
    l_mid = 0.5*(l[1:] + l[:-1])

    # FE differences are off diagonal
    df = np.diag(bar.delta_f_, k=1)

    # error estimates are off diagonal
    ddf = np.array([bar.d_delta_f_.iloc[i, i+1] for i in range(len(states)-1)])

    # Accumulate errors as sum of squares
    errors = np.array([np.sqrt((ddf[:i]**2).sum()) for i in range(len(states))])
    
    
    return l, l_mid, f, df, ddf, errors
    
def do_estimation(u_nk, method='both'):
    '''
    Run both exponential and BAR estimators and return the results in tidy dataframes.
    Arguments: u_nk in the alchemlyb format, method of fitting (String: BAR, EXP, or both)
    Returns: perWindow estimates (including errors), cumulative estimates (including errors)
    '''
    u_nk = u_nk.sort_index(level=1)
    cumulative = pd.DataFrame()
    perWindow = pd.DataFrame()
    if method=='both' or method=='BAR':
        bar = BAR()
        bar.fit(u_nk)
        ls, l_mids, fs, dfs, ddfs, errors = get_BAR(bar)
        
        cumulative[('BAR', 'f')] = fs
        cumulative[('BAR', 'errors')] = errors
        cumulative.index = ls

        perWindow[('BAR','df')] = dfs
        perWindow[('BAR', 'ddf')] = ddfs
        perWindow.index = l_mids
        
    if method=='both' or method=='EXP':
        expl, expmid, dG_fs, dG_bs = get_exponential(u_nk)

        cumulative[('EXP', 'ff')] = np.insert(np.cumsum(dG_fs),0,0)
        cumulative[('EXP', 'fb')] = np.insert(-np.cumsum(dG_bs),0,0)
        cumulative.index = expl 
        
        perWindow[('EXP','dG_f')] = dG_fs
        perWindow[('EXP','dG_b')] = dG_bs
        perWindow[('EXP', 'difference')] = np.array(dG_fs)+np.array(dG_bs)        
        perWindow.index = expmid
        
    
    perWindow.columns = pd.MultiIndex.from_tuples(perWindow.columns)
    perWindow = perWindow.fillna(0)
    cumulative.columns = pd.MultiIndex.from_tuples(cumulative.columns)
    cumulative = cumulative.fillna(0)
    
    return perWindow.copy(), cumulative.copy()
   
    
#Light-weight exponential estimator. Requires alternative parser.
def get_dG_from_data(data, temperature):
    from scipy.constants import R, calorie
    beta = 1/(R/(1000*calorie) * temperature) #So that the final result is in kcal/mol
    
    groups = data.groupby(level=0)
    dG=[]
    for name, group in groups:
        isUp = group.up
        dE = group.dE
        toAppend = [name, -1*np.log(np.mean(np.exp(-beta*dE[isUp]))), 1]
        dG.append(toAppend)
        toAppend=[name, -1*np.log(np.mean(np.exp(-beta*dE[~isUp]))), 0]
        dG.append(toAppend)
    
    dG = pd.DataFrame(dG, columns=["window", "dG", "up"])
    dG = dG.set_index('window')
    
    dG_f = dG.loc[dG.up==1] 
    dG_b = dG.loc[dG.up==0]

    dG_f = dG_f.dG.dropna()
    dG_b = dG_b.dG.dropna()

    return dG_f, dG_b