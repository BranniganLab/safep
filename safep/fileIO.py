# Import block
import numpy as np
import pandas as pd

import re

from .helpers import *


def guess_lambda(fname):
    '''
    Guess lambda based on file name (last number in the filename divided by 100). Not very good.
    Arguments: file name
    Returns: estimated lambda value
    '''
    L = int(re.findall(r'\d+', fname)[-1])/100
    return L

from pathlib import Path
def save_UNK(u_nk, filepath, safety=True):
    '''
    Write u_nk to a file
    Arguments: u_nk in the format of alchemlyb, filepath
    '''
    fpath = Path(filepath)
    if safety and fpath.is_file():
        print(f'File {filepath} already exists and safety is on. Doing nothing.')
    else:
        print(f'Writing to {filepath}')
        u_nk.to_csv(filepath)

    return
     
def read_UNK(filepath):
    '''
    Read a u_nk that was written by saveUNK.
    Arguments: filepath
    Returns: u_nk
    '''
    u_nk = pd.read_csv(filepath, index_col=[0,1], dtype=float)
    u_nk.columns = [float(x) for x in u_nk.columns]
    
    return u_nk.copy()
    

def read_FEPOUT(fileName, step=1):
    '''
    Reads all data from a NAMD fepout file unlike alchemlyb's parser.
    readFEPOUT reads each file in a single pass: keeping track of lambda values and appending each line to an array. 
    The array is cast to a dataframe at the end to avoid appending to a dataframe.
    Arguments: fileName, step (stride)
    Returns: a dataframe containing all the data in a fepout file.
    '''
    colNames = ["type",'step', 'Elec_l', 'Elec_ldl', 'vdW_l', 'vdW_ldl', 'dE', 'dE_avg', 'Temp', 'dG', 'FromLambda', "ToLambda"]

    data = []

    L = np.nan
    L2 = np.nan
    LIDWS = np.nan
    
    frame = 0
    with open(fileName) as fs:
        for line in fs:
            if line[0] == '#':
                frame = 0
                #print(line)
                Lambda = re.search('LAMBDA SET TO (\d+(\.\d+)*)', line)
                Lambda2 = re.search('LAMBDA2 (\d+(\.\d+)*)', line)
                LambdaIDWS = re.search('LAMBDA_IDWS (\d+(\.\d+)*)', line)
                if Lambda:
                    L = Lambda.group(1)
                    #print(f'L={L}')
                if Lambda2:
                    L2 = Lambda2.group(1)
                    #print(f'L2={L2}')
                if LambdaIDWS:
                    LIDWS = LambdaIDWS.group(1)
                    #print(f'LIDWS={LIDWS}')
            elif frame % step <= 1:
                if np.isnan(L):
                    print("WARNING: lambda is not defined!")
                    L = guess_lambda(fileName)
                    print("Guessing lambda to be {L} based on file name.")


                lineList = line.split()
                lineList.append(L)
                if lineList[0] == "FepEnergy:":
                    lineList.append(L2)
                elif lineList[0] == "FepE_back:":
                    lineList.append(LIDWS)
                else:
                    print(f'Unexpected line start: {lineList[0]}')
                    return 0
                data.append(lineList)
                frame = frame + 1
            else:
                frame = frame + 1

            stashL = L
            stashL2 = L2
            stashLIDWS = LIDWS

    fs.close()
    
    df = pd.DataFrame(data).dropna()
    df.columns = colNames
    df = df.iloc[:,1:].astype(float)
    df["window"]=np.mean([df.FromLambda,df.ToLambda], axis=0)
    df["up"]=df.ToLambda>df.FromLambda

    df = df.sort_index()
    return df
    
def read_Files(files, step=1):
    '''
    Batch readFEPOUT
    Arguments: file (paths), step (stride)
    Returns: a list of dataframes, one dataframe per file
    '''
    fileList = []
    for file in files:
        df = read_FEPOUT(file, step)
        fileList.append(df)
    data = pd.concat(fileList)
    
    data.index = data.window
    data["dVdW"] = data.vdW_ldl - data.vdW_l
    
    return data