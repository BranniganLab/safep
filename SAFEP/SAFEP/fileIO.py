# Import block
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1.inset_locator import inset_axes

import numpy as np
from numpy.lib.stride_tricks import sliding_window_view

from scipy.stats import linregress as lr
from scipy.stats import norm
from scipy.special import erfc
from scipy.optimize import curve_fit as scipyFit
from scipy.stats import skew

import pandas as pd

from alchemlyb.visualisation.dF_state import plot_dF_state
from alchemlyb.parsing import namd
from alchemlyb.estimators import BAR
from alchemlyb.visualisation.dF_state import plot_dF_state
from alchemlyb.visualisation import plot_convergence

import re
from tqdm import tqdm #for progress bars
from natsort import natsorted #for sorting "naturally" instead of alphabetically
from glob import glob #file regexes

from .helpers import *



#Guess lambda based on file name (last number in the filename divided by 100)
def guessLambda(fname):
    L = int(re.findall(r'\d+', fname)[-1])/100
    return L
    
def saveUNK(u_nk, filepath):
    u_nk.to_csv(filepath)
     
def readUNK(filepath):
    u_nk = pd.read_csv(filepath)
    u_nk['fep-lambda'] = u_nk['fep-lambda'].astype(str)
    u_nk = u_nk.set_index(['time', 'fep-lambda'])
    
    return u_nk.copy()
    
#redFEPOUT reads each file in a single pass: keeping track of lambda values and appending each line to an array. 
#The array is cast to a dataframe at the end to avoid appending to a dataframe
def readFEPOUT(fileName, step=1):
    colNames = ["type",'step', 'Elec_l', 'Elec_ldl', 'vdW_l', 'vdW_ldl', 'dE', 'dE_avg', 'Temp', 'dG', 'FromLambda', "ToLambda"]

    data = []

    L = np.nan
    L2 = np.nan
    LIDWS = np.nan
    
    frame = 0
    with open(fileName) as fs:
        for line in fs:
            if line[0] == '#':
                frame = 0
                #print(line)
                Lambda = re.search('LAMBDA SET TO (\d+(\.\d+)*)', line)
                Lambda2 = re.search('LAMBDA2 (\d+(\.\d+)*)', line)
                LambdaIDWS = re.search('LAMBDA_IDWS (\d+(\.\d+)*)', line)
                if Lambda:
                    L = Lambda.group(1)
                    #print(f'L={L}')
                if Lambda2:
                    L2 = Lambda2.group(1)
                    #print(f'L2={L2}')
                if LambdaIDWS:
                    LIDWS = LambdaIDWS.group(1)
                    #print(f'LIDWS={LIDWS}')
            elif frame % step <= 1:
                if np.isnan(L):
                    print("WARNING: lambda is not defined!")
                    L = guessLambda(fileName)
                    print("Guessing lambda to be {L} based on file name.")


                lineList = line.split()
                lineList.append(L)
                if lineList[0] == "FepEnergy:":
                    lineList.append(L2)
                elif lineList[0] == "FepE_back:":
                    lineList.append(LIDWS)
                else:
                    print(f'Unexpected line start: {lineList[0]}')
                    return 0
                data.append(lineList)
                frame = frame + 1
            else:
                frame = frame + 1

            stashL = L
            stashL2 = L2
            stashLIDWS = LIDWS

    fs.close()
    
    df = pd.DataFrame(data).dropna()
    df.columns = colNames
    df = df.iloc[:,1:].astype(float)
    df["window"]=np.mean([df.FromLambda,df.ToLambda], axis=0)
    df["up"]=df.ToLambda>df.FromLambda

    df = df.sort_index()
    return df
    
def readFiles(files, step=1):
    fileList = []
    for file in files:
        df = readFEPOUT(file, step)
        fileList.append(df)
    data = pd.concat(fileList)
    
    data.index = data.window
    data["dVdW"] = data.vdW_ldl - data.vdW_l
    
    return data