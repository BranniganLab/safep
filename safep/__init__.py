"""
safep

Tools for Analyzing and Debugging (SA)FEP calculations
"""

__version__ = "0.1.0"
__author__ = 'Jérôme Hénin, Ezry Santiago, Tom Joseph, Grace Brannigan'
__credits__ = 'Rutgers University - Camden'
__all__=['processing', 'plotting', 'fileIO', 'estimators', 'helpers']

from .processing import *
from .plotting import *
from .fileIO import *
from .estimators import *

