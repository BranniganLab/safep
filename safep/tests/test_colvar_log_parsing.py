from approvaltests.approvals import verify
from safep.fileIO import parse_Colvars_log
from pathlib import Path
import pytest
from collections import namedtuple

SAMPLE_RFEP_LOG = Path("../../Sample_Notebooks/RFEP_decouple.log")

@pytest.fixture
def parse_result():
    ParseResult = namedtuple('ParseReult', 'global_conf colvars biases TI_traj')
    return ParseResult(*parse_Colvars_log(SAMPLE_RFEP_LOG))

def test_global_config(parse_result):
    verify(parse_result.global_conf)

def test_colvars(parse_result):
    verify(parse_result.colvars)

def test_biases(parse_result):
    verify(parse_result.biases)

def test_TI_traj(parse_result):
    verify(parse_result.TI_traj)

