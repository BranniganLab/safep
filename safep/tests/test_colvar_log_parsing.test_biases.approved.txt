[{'key': 'harmonicwalls', 'name': 'harmonicwalls1', 'colvars': '{ d }', 'stepZeroData': 'off', 'outputEnergy': 'on', 'outputFreq': '20000', 'timeStepFactor': '1', 'writeTISamples': 'off', 'writeTIPMF': 'off', 'forceConstant': '20', 'decoupling': 'on', 'targetForceConstant': '20', 'targetNumSteps': '4000', 'targetNumStages': '25', 'outputAccumulatedWork': 'off', 'targetEquilSteps': '500', 'lambdaSchedule': '', 'lambdaExponent': '4', 'lowerWalls': '{ 0 }', 'upperWalls': '{ 5 }', 'upperWallConstant': '20'}]
